### PR TITLE
Make fake client actions use fully qualified resource

### PR DESCRIPTION
**Pull request by caesarxuchao**
The output of a versioned clientset is version object. The fake client used to assume only internal objects will be returned. This PR removes this assumption by making fake actions initialized with a fully qualified resource instead of a resource string.

We have to regenerate fake clients in release_1_2 clientset to let it compile. For the test fakes, we are breaking the backwards compatibility promise.

Part of #24155.

### CODE REVIEW

**Comment by caesarxuchao** [comment]
We have to modify the release_1_2 fake clients, otherwise it won't compile.

**Comment by lavalamp** [comment]
@deads2k can you review this? I know you're a big proponent of making clear distinctions between resources and kinds. :)

**Comment by deads2k** [comment]
> @deads2k can you review this? I know you're a big proponent of making clear distinctions between resources and kinds. :)

I like the API change a lot.  That LGTM.  I'd like a follow-on for dealing with `Matches`, but I wouldn't hold this up on that.

I don't have the background to review the generator changes themselves though.

@krousey  maybe?

File: cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go [hunk 1]
```diff
@@ -78,16 +79,31 @@ func (g *genFakeForType) GenerateType(c *generator.Context, t *types.Type, w io.
 	pkg := filepath.Base(t.Name.Package)
 	const pkgTestingCore = "k8s.io/kubernetes/pkg/client/testing/core"
 	namespaced := !(types.ExtractCommentTags("+", t.SecondClosestCommentLines)["nonNamespaced"] == "true")
+	canonizeGroup := func(group string) string {
```
Comment (krousey) [inline cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go]: I think you mean canonicalize.

https://en.wikipedia.org/wiki/Canonization_(disambiguation)

> Sometimes, the term is used incorrectly to refer to canonicalization, i.e., finding a canonical form

File: cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go [hunk 2]
```diff
@@ -78,16 +79,31 @@ func (g *genFakeForType) GenerateType(c *generator.Context, t *types.Type, w io.
...
-		"watchInterface":   c.Universe.Type(types.Name{Package: "k8s.io/kubernetes/pkg/watch", Name: "Interface"}),
-		"apiDeleteOptions": c.Universe.Type(types.Name{Package: "k8s.io/kubernetes/pkg/api", Name: "DeleteOptions"}),
-		"apiListOptions":   c.Universe.Type(types.Name{Package: "k8s.io/kubernetes/pkg/api", Name: "ListOptions"}),
-		"Everything":       c.Universe.Function(types.Name{Package: "k8s.io/kubernetes/pkg/labels", Name: "Everything"}),
+		"type":                 t,
+		"package":              pkg,
+		"Package":              namer.IC(pkg),
+		"namespaced":           namespaced,
+		"Group":                namer.IC(g.group),
+		"group":                canonizeGroup(g.group),
```
Comment (krousey) [inline cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go]: Why did you write anonymous functions if it's only going to be called once. Why not just use a variable?

```go
canonicalGroup := g.group
if canonicalGroup == "core" {
	canonicalGroup = ""
}

...

"group": canonicalGroup
```

File: pkg/client/clientset_generated/release_1_2/typed/extensions/v1beta1/fake/fake_job.go [hunk 3]
```diff
@@ -30,9 +31,11 @@ type FakeJobs struct {
 	ns   string
 }
 
+const resource = unversioned.GroupVersionResource{group: "extensions", version: "v1beta1", resource: "job"}
```
Comment (krousey) [inline pkg/client/clientset_generated/release_1_2/typed/extensions/v1beta1/fake/fake_job.go]: Does this compile? I don't think you can have struct constants.

File: pkg/client/testing/core/fixture.go [hunk 4]
```diff
@@ -57,7 +57,10 @@ type ObjectScheme interface {
 // TODO: add support for sub resources
 func ObjectReaction(o ObjectRetriever, mapper meta.RESTMapper) ReactionFunc {
 	return func(action Action) (bool, runtime.Object, error) {
-		kind, err := mapper.KindFor(action.GetResource())
+		resource := action.GetResource()
+		kind, err := mapper.KindFor(resource)
+		// This is a temporary fix. Because there is no internal resource, so the caller has no way to express it expects to get an internal kind back.
```
Comment (krousey) [inline pkg/client/testing/core/fixture.go]: What is the eventual solution to this? So others who read this know.

File: pkg/client/clientset_generated/release_1_3/typed/core/v1/fake/fake_event_expansion.go [hunk 5]
```diff
@@ -18,16 +18,17 @@ package fake
...
+	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/api/v1"
 	"k8s.io/kubernetes/pkg/client/testing/core"
 	"k8s.io/kubernetes/pkg/fields"
 	"k8s.io/kubernetes/pkg/runtime"
 )
 
 func (c *FakeEvents) CreateWithEventNamespace(event *v1.Event) (*v1.Event, error) {
-	action := core.NewRootCreateAction("events", event)
+	action := core.NewRootCreateAction(unversioned.GroupVersionResource{Version: "v1", Resource: "events"}, event)
```
Comment (krousey) [inline pkg/client/clientset_generated/release_1_3/typed/core/v1/fake/fake_event_expansion.go]: eventsResource?

File: pkg/client/clientset_generated/release_1_3/typed/core/v1/fake/fake_namespace_expansion.go [hunk 6]
```diff
@@ -17,14 +17,15 @@ limitations under the License.
...
+	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/api/v1"
 	"k8s.io/kubernetes/pkg/client/testing/core"
 )
 
 func (c *FakeNamespaces) Finalize(namespace *v1.Namespace) (*v1.Namespace, error) {
 	action := core.CreateActionImpl{}
 	action.Verb = "create"
-	action.Resource = "namespaces"
+	action.Resource = unversioned.GroupVersionResource{Version: "v1", Resource: "namespaces"}
```
Comment (krousey) [inline pkg/client/clientset_generated/release_1_3/typed/core/v1/fake/fake_namespace_expansion.go]: namespacesResource?

File: pkg/client/clientset_generated/release_1_3/typed/core/v1/fake/fake_pod_expansion.go [hunk 7]
```diff
@@ -25,7 +26,7 @@ import (
 func (c *FakePods) Bind(binding *v1.Binding) error {
 	action := core.CreateActionImpl{}
 	action.Verb = "create"
-	action.Resource = "pods"
+	action.Resource = unversioned.GroupVersionResource{Version: "v1", Resource: "pods"}
```
Comment (krousey) [inline pkg/client/clientset_generated/release_1_3/typed/core/v1/fake/fake_pod_expansion.go]: podsResource?

File: pkg/client/clientset_generated/release_1_3/typed/core/v1/fake/fake_service_expansion.go [hunk 8]
```diff
@@ -17,10 +17,11 @@ limitations under the License.
...
 
 import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/client/restclient"
 	"k8s.io/kubernetes/pkg/client/testing/core"
 )
 
 func (c *FakeServices) ProxyGet(scheme, name, port, path string, params map[string]string) restclient.ResponseWrapper {
-	return c.Fake.InvokesProxy(core.NewProxyGetAction("services", c.ns, scheme, name, port, path, params))
+	return c.Fake.InvokesProxy(core.NewProxyGetAction(unversioned.GroupVersionResource{Version: "v1", Resource: "services"}, c.ns, scheme, name, port, path, params))
```
Comment (krousey) [inline pkg/client/clientset_generated/release_1_3/typed/core/v1/fake/fake_service_expansion.go]: servicesResource?

File: pkg/client/clientset_generated/release_1_3/typed/extensions/v1beta1/fake/fake_deployment_expansion.go [hunk 9]
```diff
@@ -17,14 +17,15 @@ limitations under the License.
...
+	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/apis/extensions/v1beta1"
 	"k8s.io/kubernetes/pkg/client/testing/core"
 )
 
 func (c *FakeDeployments) Rollback(deploymentRollback *v1beta1.DeploymentRollback) error {
 	action := core.CreateActionImpl{}
 	action.Verb = "create"
-	action.Resource = "deployments"
+	action.Resource = unversioned.GroupVersionResource{Group: "extensions", Version: "v1beta1", Resource: "deployments"}
```
Comment (krousey) [inline pkg/client/clientset_generated/release_1_3/typed/extensions/v1beta1/fake/fake_deployment_expansion.go]: deploymentsResource?

File: pkg/client/clientset_generated/release_1_2/typed/core/v1/fake/fake_event_expansion.go [hunk 10]
```diff
@@ -18,16 +18,17 @@ package fake
...
+	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/api/v1"
 	"k8s.io/kubernetes/pkg/client/testing/core"
 	"k8s.io/kubernetes/pkg/fields"
 	"k8s.io/kubernetes/pkg/runtime"
 )
 
 func (c *FakeEvents) CreateWithEventNamespace(event *v1.Event) (*v1.Event, error) {
-	action := core.NewRootCreateAction("events", event)
+	action := core.NewRootCreateAction(unversioned.GroupVersionResource{Version: "v1", Resource: "events"}, event)
```
Comment (krousey) [inline pkg/client/clientset_generated/release_1_2/typed/core/v1/fake/fake_event_expansion.go]: eventsResource?

File: pkg/client/clientset_generated/release_1_2/typed/core/v1/fake/fake_namespace_expansion.go [hunk 11]
```diff
@@ -17,14 +17,15 @@ limitations under the License.
...
+	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/api/v1"
 	"k8s.io/kubernetes/pkg/client/testing/core"
 )
 
 func (c *FakeNamespaces) Finalize(namespace *v1.Namespace) (*v1.Namespace, error) {
 	action := core.CreateActionImpl{}
 	action.Verb = "create"
-	action.Resource = "namespaces"
+	action.Resource = unversioned.GroupVersionResource{Version: "v1", Resource: "namespaces"}
```
Comment (krousey) [inline pkg/client/clientset_generated/release_1_2/typed/core/v1/fake/fake_namespace_expansion.go]: namespacesResource?

File: pkg/client/clientset_generated/release_1_2/typed/core/v1/fake/fake_pod_expansion.go [hunk 12]
```diff
@@ -25,7 +26,7 @@ import (
 func (c *FakePods) Bind(binding *v1.Binding) error {
 	action := core.CreateActionImpl{}
 	action.Verb = "create"
-	action.Resource = "pods"
+	action.Resource = unversioned.GroupVersionResource{Version: "v1", Resource: "pods"}
```
Comment (krousey) [inline pkg/client/clientset_generated/release_1_2/typed/core/v1/fake/fake_pod_expansion.go]: podsResource?

File: pkg/client/clientset_generated/release_1_2/typed/core/v1/fake/fake_service_expansion.go [hunk 13]
```diff
@@ -17,10 +17,11 @@ limitations under the License.
...
 
 import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/client/restclient"
 	"k8s.io/kubernetes/pkg/client/testing/core"
 )
 
 func (c *FakeServices) ProxyGet(scheme, name, port, path string, params map[string]string) restclient.ResponseWrapper {
-	return c.Fake.InvokesProxy(core.NewProxyGetAction("services", c.ns, scheme, name, port, path, params))
+	return c.Fake.InvokesProxy(core.NewProxyGetAction(unversioned.GroupVersionResource{Version: "v1", Resource: "services"}, c.ns, scheme, name, port, path, params))
```
Comment (krousey) [inline pkg/client/clientset_generated/release_1_2/typed/core/v1/fake/fake_service_expansion.go]: servicesResource?

File: pkg/client/clientset_generated/release_1_2/typed/extensions/v1beta1/fake/fake_deployment_expansion.go [hunk 14]
```diff
@@ -17,14 +17,15 @@ limitations under the License.
...
+	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/apis/extensions/v1beta1"
 	"k8s.io/kubernetes/pkg/client/testing/core"
 )
 
 func (c *FakeDeployments) Rollback(deploymentRollback *v1beta1.DeploymentRollback) error {
 	action := core.CreateActionImpl{}
 	action.Verb = "create"
-	action.Resource = "deployments"
+	action.Resource = unversioned.GroupVersionResource{Group: "extensions", Version: "v1beta1", Resource: "deployments"}
```
Comment (krousey) [inline pkg/client/clientset_generated/release_1_2/typed/extensions/v1beta1/fake/fake_deployment_expansion.go]: deploymentsResource?

File: pkg/client/typed/generated/core/unversioned/fake/fake_event_expansion.go [hunk 15]
```diff
@@ -18,15 +18,16 @@ package fake
...
 	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/client/testing/core"
 	"k8s.io/kubernetes/pkg/fields"
 	"k8s.io/kubernetes/pkg/runtime"
 )
 
 func (c *FakeEvents) CreateWithEventNamespace(event *api.Event) (*api.Event, error) {
-	action := core.NewRootCreateAction("events", event)
+	action := core.NewRootCreateAction(unversioned.GroupVersionResource{Resource: "events"}, event)
```
Comment (krousey) [inline pkg/client/typed/generated/core/unversioned/fake/fake_event_expansion.go]: eventsResource?

File: cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go (hunk 1 above)
Comment (caesarxuchao) [inline cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go]: Thanks.

File: pkg/client/clientset_generated/release_1_3/typed/core/v1/fake/fake_event_expansion.go (hunk 5 above)
Comment (caesarxuchao) [inline pkg/client/clientset_generated/release_1_3/typed/core/v1/fake/fake_event_expansion.go]: Thanks. This change is made by a script. I'll update it.

File: cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go (hunk 2 above)
Comment (caesarxuchao) [inline cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go]: Thanks. Done.

File: pkg/client/clientset_generated/release_1_2/typed/extensions/v1beta1/fake/fake_job.go (hunk 3 above)
Comment (caesarxuchao) [inline pkg/client/clientset_generated/release_1_2/typed/extensions/v1beta1/fake/fake_job.go]: Sorry that was in an intermediate commit. It's fixed by other commits.

File: pkg/client/testing/core/fixture.go [hunk 16]
```diff
@@ -57,7 +57,10 @@ type ObjectScheme interface {
 // TODO: add support for sub resources
 func ObjectReaction(o ObjectRetriever, mapper meta.RESTMapper) ReactionFunc {
 	return func(action Action) (bool, runtime.Object, error) {
-		kind, err := mapper.KindFor(unversioned.GroupVersionResource{Resource: action.GetResource()})
+		resource := action.GetResource()
+		kind, err := mapper.KindFor(resource)
+		// This is a temporary fix. Because there is no internal resource, so the caller has no way to express that it expects to get an internal kind back. A more proper fix will be directly specify the Kind when build the action.
```
Comment (caesarxuchao) [inline pkg/client/testing/core/fixture.go]: I added a sentence on what's the proper fix in my mind, but I don't think we'll have time to implement it any soon.

**Comment by caesarxuchao** [comment]
Comments addressed. PTAL. Thanks!

File: pkg/client/testing/core/fixture.go (hunk 16 above)
Comment (krousey) [inline pkg/client/testing/core/fixture.go]: Thanks. It's a bit of a long line.

**Comment by krousey** [comment]
One minor comment about line length, otherwise LGTM from me.

**Comment by caesarxuchao** [comment]
Thanks. I'll fix that long line and squash.

**Comment by caesarxuchao** [comment]
Fixed and squashed. I applied the label. Thanks.

**Comment by caesarxuchao** [comment]
@k8s-bot unit test this, issue #23361

**Comment by caesarxuchao** [comment]
@k8s-bot unit test this, issue #23361

**Comment by caesarxuchao** [comment]
@k8s-bot unit test this, issue #23361

**Comment by caesarxuchao** [comment]
verify-govet.sh failed:
```
Verifying ./hack/../hack/verify-govet.sh
pkg/client/clientset_generated/release_1_2/fake/clientset_generated.go:59: k8s.io/kubernetes/pkg/client/typed/discovery/fake.FakeDiscovery composite literal uses unkeyed fields
pkg/client/clientset_generated/release_1_2/fake/clientset_generated.go:66: k8s.io/kubernetes/pkg/client/clientset_generated/release_1_2/typed/core/v1/fake.FakeCore composite literal uses unkeyed fields
pkg/client/clientset_generated/release_1_2/fake/clientset_generated.go:71: k8s.io/kubernetes/pkg/client/clientset_generated/release_1_2/typed/extensions/v1beta1/fake.FakeExtensions composite literal uses unkeyed fields
FAILED   ./hack/../hack/verify-govet.sh	102s
```

but it doesn't fail when I run it locally. My go version is 1.6.1. @goltermann do you know what's happening?

log: https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/24398/kubernetes-pull-test-unit-integration/22136/
the go vet offender: https://github.com/kubernetes/kubernetes/pull/24398/files#diff-dcdb76ec058e5f683dcc2563498906d2R59

**Comment by caesarxuchao** [comment]
@k8s-bot test this #24579

**Comment by caesarxuchao** [comment]
@k8s-bot unit test this, issue #24579

**Comment by goltermann** [comment]
@caesarxuchao I did see this in one case when making fixes.  @krousey helped out.  The way to expose this locally is:

<from kub root>
```
make -C hack/jenkins/test-image container
WORKSPACE=<some tmp dir> hack/jenkins/gotest-dockerized.sh
```

What's not clear to me is why the containerized test script checks more files - I couldn't find any functional difference in what was being run and tested.

**Comment by caesarxuchao** [comment]
@k8s-bot unit test this, issue #24579

**Comment by caesarxuchao** [comment]
Thanks @goltermann.

**Pull request merged**